src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -163,146 +162,124 @@ public void Clear()
+                        EvictDanglingEntries();
                     }
 
-                    Debug.Assert(cache.Count < MaxTrackedContexts);
-
-                    // Use a defensive copy of the options instance as key to
-                    // avoid capturing references to any caching contexts.
-                    var key = new JsonSerializerOptions(options);
-                    Debug.Assert(key._cachingContext == null);
-
-                    ctx = new CachingContext(options);
-                    bool success = cache.TryAdd(key, new WeakReference<CachingContext>(ctx));
-                    Debug.Assert(success);
+                    if (s_size == MaxTrackedContexts)
+                    {
+                        // Cache is full; return a fresh instance.
+                        return new CachingContext(options);
+                    }
 
+                    var ctx = new CachingContext(options);
+                    s_trackedContexts[s_size++] = new WeakReference<CachingContext>(ctx);
                     return ctx;
                 }
             }
 
-            public static void Clear()
+            private static bool TryGetSharedContext(
+                JsonSerializerOptions options,
+                out bool foundDanglingEntries,
+                [NotNullWhen(true)] out CachingContext? result)
             {
-                lock (s_cache)
-                {
-                    s_cache.Clear();
-                    s_recentEvictionCounts.Clear();
-                    s_evictionRunsToSkip = 0;
-                }
-            }
 
-            private static bool TryEvictDanglingEntries()
-            {
-                // Worst case scenario, the cache has been filled with permanent entries.
-                // Evictions are synchronized and each run is in the order of microseconds,
-                // so we want to avoid triggering runs every time an instance is initialized,
-                // For this reason we use a backoff strategy to average out the cost of eviction
-                // across multiple initializations. The backoff count is determined by the eviction
-                // rates of the most recent runs.
-
-                Debug.Assert(Monitor.IsEntered(s_cache));
-
-                if (s_evictionRunsToSkip > 0)
-                {
-                    --s_evictionRunsToSkip;
-                    return false;
-                }
-
-                int currentEvictions = 0;
-                foreach (KeyValuePair<JsonSerializerOptions, WeakReference<CachingContext>> kvp in s_cache)
-                {
-                    if (!kvp.Value.TryGetTarget(out _))
-                    {
-                        bool result = s_cache.TryRemove(kvp.Key, out _);
-                        Debug.Assert(result);
-                        currentEvictions++;
-                    }
-                }
+                WeakReference<CachingContext>?[] trackedContexts = s_trackedContexts;
+                int size = s_size;
```

Comment:
This is accessing shared mutable state without a lock.  At a minimum we should have a comment explaining why that's ok.